File: src/main/java/core/procedures/meta/WIthMeta.java
package core.procedures.meta;

import core.exceptions.WrongTypeException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.IMeta;
import core.scm.SCMSymbol;

import java.util.Map;

public final class WIthMeta extends AFn {

  public WIthMeta() {
    super(new FnArgsBuilder().minArgs(2).maxArgs(2).mandatoryArgsTypes(new Class[] {IMeta.class, Map.class}));
  }

  @Override
  public boolean isPure() {
    return false;
  }

  @Override
  public String getName() {
    return "with-meta";
  }

  @Override
  public Object apply2(Object obj, Object meta) {
    if (obj instanceof SCMSymbol) {
      return new SCMSymbol(((SCMSymbol) obj).getName(), (Map)meta);
    }
    throw new WrongTypeException(getName(), "IMeta", obj);
  }
}

File: src/main/java/core/procedures/meta/WIthMeta.kt
package core.procedures.meta

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.IMeta
import core.scm.Symbol

class WIthMeta : AFn<Any?, Symbol>(name = "with-meta", arity = Exactly(2),
                                   mandatoryArgsTypes = arrayOf(IMeta::class.java, Map::class.java)) {

    override operator fun invoke(arg1: Any?, arg2: Any?) = Symbol((arg1 as Symbol).name, arg2 as Map<*, *>?)
}
